File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> st;
        int left=0;
        int longest=0;
        for(int right=0; right<s.length();right++){
            
            // check is right is in the set 
            while(st.find(s[right])!=st.end()){
                st.erase(s[left]);
                left++;
            } 
            //check the size of substring 
            //right-left is alway be the length of the window 
            longest=max(longest,right-left+1);
            st.insert(s[right]);

        }

    return longest;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        st=set() 
        left=0
        longest=0

        for right in range(len(s)):

            while s[right] in st:
                st.remove(s[left])
                left+=1

            longest=max(right-left+1,longest)
            st.add(s[right])

        return longest